### ISSUE TITLE
compile  error: nonnull argument 'pk_unpack_info' compared to NULL [-Werror=nonnull-compare]

### ISSUE THREAD

**Issue by grooverdan** [facebook/mysql-5.6#346]
On Ubuntu-16.04

<pre>
add-apt-repository -y ppa:ubuntu-toolchain-r/test
apt-get update && apt-get install -y gcc-6 g++-6

08:22:59 + cmake /source -DCMAKE_BUILD_TYPE=Debug -DBUILD_CONFIG=mysql_release -DWITH_PROTOBUF=system -DWITH_NDB_JAVA=NO -DCMAKE_C_COMPILER=gcc-6 -DCMAKE_CXX_COMPILER=g++-6 '-DCMAKE_C_FLAGS_RELWITHDEBINFO=-m64 -O3 -g -mtune=native' '-DCMAKE_CXX_FLAGS_RELWITHDEBINFO=-m64 -O3 -g -mtune=native' '-DCMAKE_C_FLAGS=-m64 -O3 -g -mtune=native' '-DCMAKE_CXX_FLAGS=-m64 -O3 -g -mtune=native' -DWITH_SSL=system -DCMAKE_INSTALL_PREFIX=/mysql/facebook-upstream-Debug-webscalesql-5.6.27.75 -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
08:22:59 -- Running cmake version 3.5.1
08:22:59 -- Found Git: /usr/bin/git (found version "2.7.4") 
08:23:00 -- The C compiler identification is GNU 6.2.0
08:23:01 -- The CXX compiler identification is GNU 6.2.0
08:23:01 -- Check for working C compiler: /usr/bin/gcc-6
08:23:02 -- Check for working C compiler: /usr/bin/gcc-6 -- works
08:23:02 -- Detecting C compiler ABI info
08:23:04 -- Detecting C compiler ABI info - done
08:23:05 -- Detecting C compile features
08:23:07 -- Detecting C compile features - done
08:23:07 -- Check for working CXX compiler: /usr/bin/g++-6
08:23:07 -- Check for working CXX compiler: /usr/bin/g++-6 -- works
...
...
08:25:33 [ 16%] Building CXX object storage/rocksdb/CMakeFiles/rocksdb_se.dir/ha_rocksdb.cc.o
08:25:33 cd /build/build/storage/rocksdb && ccache /usr/bin/g++-6   -DHAVE_CONFIG_H -DOS_LINUX -DROCKSDB_LIB_IO_POSIX -DROCKSDB_PLATFORM_POSIX -DZLIB -I/build/build/include -I/source/rocksdb -I/source/rocksdb/include -I/source/rocksdb/third-party/gtest-1.7.0/fused-src -I/source/include -I/source/sql -I/source/regex -I/source/zlib  -m64 -O3 -g -mtune=native -Wall -Wextra -Wformat-security -Woverloaded-virtual -Wno-unused-parameter -Werror -DFORCE_INIT_OF_VARS -std=c++11 -g -fabi-version=2 -fno-omit-frame-pointer -fno-strict-aliasing -DENABLED_DEBUG_SYNC -DSAFE_MUTEX   -o CMakeFiles/rocksdb_se.dir/ha_rocksdb.cc.o -c /source/storage/rocksdb/ha_rocksdb.cc
08:25:33 /source/storage/rocksdb/ha_rocksdb.cc: In member function 'void myrocks::ha_rocksdb::convert_record_to_storage_format(const rocksdb::Slice&, myrocks::Rdb_string_writer*, rocksdb::Slice*)':
08:25:33 /source/storage/rocksdb/ha_rocksdb.cc:4057:25: error: nonnull argument 'pk_unpack_info' compared to NULL [-Werror=nonnull-compare]
08:25:33    DBUG_ASSERT_IMP(m_maybe_unpack_info, pk_unpack_info);
</pre>

**Comment by gunnarku** [comment]
Thank you for reporting the compilation issues!

Internally we're compiling MyRocks with GCC 6.x.x. with `-DMYSQL_MAINTAINER_MODE=0`. Please add `-DMYSQL_MAINTAINER_MODE=0` to your CMake command-line, this should unblock you.

**Comment by jkedgar** [comment]
Possible solution: https://reviews.facebook.net/D65307

**Comment by grooverdan** [comment]
Sorry for being slow in testing:

Works great on gcc-6 and clang versions. Fails on gcc-5.5 however.

<pre>
10:31:55 + cmake /source -DCMAKE_BUILD_TYPE=Debug -DBUILD_CONFIG=mysql_release -DWITH_PROTOBUF=system -DWITH_NDB_JAVA=NO -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ '-DCMAKE_C_FLAGS_RELWITHDEBINFO=-m64 -O3 -g -mtune=native' '-DCMAKE_CXX_FLAGS_RELWITHDEBINFO=-m64 -O3 -g -mtune=native' '-DCMAKE_C_FLAGS=-m64 -O3 -g -mtune=native' '-DCMAKE_CXX_FLAGS=-m64 -O3 -g -mtune=native' -DWITH_SSL=system -DCMAKE_INSTALL_PREFIX=/mysql/facebook-develop-Debug-gcc-6-D65307 -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
10:31:55 -- Running cmake version 3.5.1
10:31:55 -- Found Git: /usr/bin/git (found version "2.7.4") 
10:31:55 -- The C compiler identification is GNU 5.4.0
10:31:55 -- The CXX compiler identification is GNU 5.4.0
10:36:11 [ 20%] Building CXX object storage/rocksdb/CMakeFiles/rocksdb_se_embedded.dir/ha_rocksdb.cc.o
10:36:11 cd /build/build/storage/rocksdb && ccache /usr/bin/g++   -DHAVE_CONFIG_H -DHAVE_SCHED_GETCPU=1 -DOS_LINUX -DROCKSDB_LIB_IO_POSIX -DROCKSDB_PLATFORM_POSIX -DZLIB -I/build/build/include -I/source/rocksdb -I/source/rocksdb/include -I/source/rocksdb/third-party/gtest-1.7.0/fused-src -I/source/include -I/source/sql -I/source/regex -I/source/zlib  -m64 -O3 -g -mtune=native -Wall -Wextra -Wformat-security -Woverloaded-virtual -Wno-unused-parameter -Wno-nonnull-compare -Werror -DFORCE_INIT_OF_VARS -std=c++11 -Wnonnull-compare -g -fabi-version=2 -fno-omit-frame-pointer -fno-strict-aliasing -DENABLED_DEBUG_SYNC -DSAFE_MUTEX   -fPIC -o CMakeFiles/rocksdb_se_embedded.dir/ha_rocksdb.cc.o -c /source/storage/rocksdb/ha_rocksdb.cc
10:36:11 g++: error: unrecognized command line option '-Wnonnull-compare'
</pre>

**Comment by grooverdan** [comment]
thanks @gunnarku  for the workaround too :cake:

**Comment by dutow** [comment]
I noticed the same problem at multiple locations when trying to compile myrocks with clang and maintainer mode for percona-server. I can't see that reviews page, but I found three possible solutions/workarounds so far:

* removing the non null attributes
* removing the assertions
* disabling the warning for the affected files

I think the first one is the best, because:
* the second removes currently existing assertions
* the third works for normal debug builds, but it doesn't work in sanitized (asan, msan, ...) builds, as those are compiled with -O1, and clang optimizes away the check

What's your (preferred) workaround/solution in that review?

**Comment by laurynas-biveinis** [comment]
Unless I'm missing something, I'd suggest removing nonnull attributes. They silently optimize away any runtime NULL checks, including defensive ones, while not providing enough protection against actual nullptr being passed.

**Comment by george-lorch** [comment]
Agreed with @laurynas-biveinis nonnull attribute is a simple compile time check only and does not at all eliminate runtime issues. IMHO the use of this attribute over proper runtime checks should be abolished entirely.
